migrations/dimensions-config/1fae088c80b6.py

```diff
@@ -0,0 +1,877 @@
+        "exposure",
+        ["instrument", "group"],
+        schema=ctx.schema,
+    )
+
+
+def _migrate_day_obs(ctx: _Context) -> None:
+    # Before doing anything else, generate the rows for the new day_obs table
+    # from the data in the exposure table.  This is prone to failure due to the
+    # need to import instrument classes.
+    _LOG.info("Generating data for day_obs table from exposure_table")
+    day_obs_rows = list(_generate_day_obs_rows(ctx))
+
+    # Create day_obs table
+    _LOG.info("Creating day_obs table")
+    timespan_columns: list[sa.Column]
+    if ctx.is_postgres:
+        # Postgres uses a non-standard range datatype for representing timespans.
+        timespan_columns = [sa.Column("timespan", INT8RANGE)]
+    else:
+        timespan_columns = [
+            sa.Column("timespan_begin", sa.BigInteger),
+            sa.Column("timespan_end", sa.BigInteger),
+        ]
+
+    check_constraints = []
+    if ctx.is_sqlite:
+        check_constraints = [_make_string_check_constraint("instrument", 32, "day_obs_len_instrument")]
+
+    table = op.create_table(
+        "day_obs",
+        sa.Column("instrument", sa.String(32), primary_key=True),
+        sa.Column("id", sa.BigInteger, primary_key=True, autoincrement=False),
+        *timespan_columns,
+        sa.schema.ForeignKeyConstraint(
+            columns=["instrument"],
+            refcolumns=["instrument.name"],
+            name="fkey_day_obs_instrument_name_instrument",
+        ),
+        *check_constraints,
+        schema=ctx.schema,
+    )
+
+    # Populate the day_obs table based on the data in the exposure table.
+    _LOG.info("Populating day_obs table")
+    op.bulk_insert(table, day_obs_rows)
+
+    # Create index on instrument
+    _LOG.info("Creating instrument index for day_obs table")
+    op.create_index(
+        "day_obs_fkidx_instrument",
+        "day_obs",
+        ["instrument"],
+        schema=ctx.schema,
+    )
+
+    # Update exposure table to reference day_obs table
+    _LOG.info("Updating exposure table to reference day_obs table")
+    with op.batch_alter_table("exposure", schema=ctx.schema) as batch_op:
+        batch_op.alter_column("day_obs", nullable=False)
+        batch_op.create_foreign_key(
+            constraint_name="fkey_exposure_day_obs_instrument_id_instrument_day_obs",
+            referent_table="day_obs",
+            local_cols=["instrument", "day_obs"],
+            remote_cols=["instrument", "id"],
+            referent_schema=ctx.schema,
+        )
+
+    # Create index on exposure for day_obs fkey
+    op.create_index(
+        "exposure_fkidx_instrument_day_obs",
+        "exposure",
+        ["instrument", "day_obs"],
+        schema=ctx.schema,
+    )
+
+
+def _migrate_dimensions_json(ctx: _Context) -> None:
+    _LOG.info("Updating dimensions.json in ButlerAttributes")
+    attributes = ButlerAttributes(ctx.bind, ctx.schema)
+    attributes.update("config:dimensions.json", _dimensions_json_universe_6)
```

Comment:
I prefer read-modify-save style for "config:dimensions.json" - easier to read compared to full JSON dump and also could make it easier to reuse for different namespace name if we ever need to do it. I'd like to think that v5 config is identical in all our repos, but there is no guarantee that it actually is, so I prefer to only update parts that actually change.

Reply:
if the version 5 config we are upgrading is not really a version 5 config, don't we have some problems?

Should we be calculating the MD5 for these universe versions and warning if v5 isn't really v5?

Reply:
Well, we may or may not have problem, depending on who "we" are 🙂 If someone out there decides to do something crazy with their database we cannot stop them. We can try and verify that it is indeed v5 during migration, but maybe it's easier to only change what we really need to change.

Reply:
Before we had versioning we had no choice but to apply diffs to the universe because we had no idea which universe we were starting from. I worry that allowing versioned universes to diverge over time as we apply diffs could lead to a situation where two repos have version 50 but have wildly different JSON files.

Reply:
I can understand that argument, still there are other consideration here. I'm not very happy to see the whole content of that thing in daf_butler_migrate. Potentially, if we could keep those reference versions in daf_butler, as you suggested on Slack already, I could be more agreeable if we could read them from there (and convert to JSON). For me it's mostly philosophical difference - I prefer to change as little as possible in the existing system, if I can help it 🙂 The use case that I'm pondering - imagine that someone discovers a bug and we can fix it only by making an emergency patch to dimensions.json, but we cannot do it using regular migration tools (because for some reason people cannot migrate those affected repos to latest version). If we later overwrite things unconditionally the patch will be lost. This situation can probably be solved with the branches in the migration scripts but branches are very hard to maintain.

Reply:
I think my plan for this particular migration will be:
* Store the historical dimensions.yaml in daf_butler
* Before starting the migration, verify that the dimensions.json in the DB matches the expected v5 universe.  If it doesn't match, raise an exception so we become aware of the mismatch and can decide what to do.  As far as we know they all "should" match, so if they don't it would be good to know why.
* Write the v6 universe as a whole to the DB.  In my opinion patching it manually is unnecessary work and is error-prone.  I caused a bug at my last job by patching JSON during a migration but accidentally applying a property at the wrong level of nesting, which then caused problems in a later migration.  If the user has modified the universe, there's no guarantee that patching it will generate a correct universe anyway. 

I think in the face of user modifications we need to decide what to do on a case-by-case basis, so detecting and reporting the situation is the right thing to do.